### ISSUE TITLE
Support `push --force-with-lease`

### ISSUE THREAD

**Issue by sonicdoe** [git-up/GitUp#333]
See Atlassian’s “[--force considered harmful; understanding git's --force-with-lease](https://developer.atlassian.com/blog/2015/04/force-with-lease/)” article (or [thoughtbot’s article](https://robots.thoughtbot.com/git-push-force-with-lease) for a shorter version) for why `git push --force-with-lease` is a good idea.

It would be great if GitUp added support for this. It might even make sense to make this the default and present a warning dialog if someone updated the branch upstream.

**Comment by swisspol** [comment]
Unless I'm missing something at a glance, this doesn't seem to be a git primitive so I'd be surprised if libgit2 supports it. Probably needs to be "emulated" on top of it.

**Comment by QueenC** [comment]
Envie de rejoindre une startup tech à taille humaine sur les sujets Data, Automation, Performance et Community?#symfony_jobs#php#job
http://bit.ly/2xyPF7W

**Comment by schwa** [comment]
It's churlish to leave a +1. But um +1.

This would be a great feature to add for safety.

**Comment by sesam** [comment]
Yeah, if this isn't a server-side feature then it basically just does `git fetch` and checks that the branch to-be-pished didn't get anything new meanwhile (but since when?). As gitup normally does `git fetch` every few minutes this would be tricky. Likely there should be some commit hash used as base and whenever remote stacks any new commits on top of the "remembered" hash that would be considered not safe to push over? Or maybe just always complain about force pushing over a remote branch that would mean throwing away some commits on the remote -- and the list of those commits could be shown in a dialog before the user presses on Force overwrite.

**Comment by jlukas** [comment]
What about supporting something like [`--force-if-includes`](https://git-scm.com/docs/git-push#Documentation/git-push.txt---no-force-if-includes)?

> This option enables a check that verifies if the tip of the remote-tracking ref is reachable from one of the "reflog" entries of the local branch based in it for a rewrite.

That would avoid the issue of automatic fetches affecting the behaviour of force pushes.
